### ISSUE TITLE
[bug] jump drive equipped player fleet lands to refuel with <200 fuel

### ISSUE THREAD

**Issue by LocalGod79** [endless-sky/endless-sky#1238]
even if their fuel is >100, they also have a hyperdrive, and their next jump is _by_ hyperdrive into a habited system.

it's a very niche problem, but it is unexpected and possibly annoying - especially if that last jump is into an unexpected war-zone, and the player was expecting their similarly equipped fleet to jump with them.

edit; was reported on Steam, is easy to reproduce.

**Comment by Loke85** [comment]
Not topical, but it was I who reported this on steam.
Is it such that github>google groups>steam? So this is the primary place?

**Comment by LocalGod79** [comment]
M.Z. as well as other involved community members check all three regularly: GitHub most frequently; and Google Groups facilitates file transfers (e.g. pilot save files).

Steam just happens to be the place that seems most accessible to most E.S. users.

**Comment by Alkallid** [comment]
Occasionally people can use Dropbox or whatever in Steam. Also, yeah, most users contribute to those forums, but such that it clumps into spam.

**Comment by LocalGod79** [comment]
having had some time to mull this over in my subconscious, i believe this is due to the a.i. behaviour of reserving an "emergency jump"'s worth of fuel (which is not a bad thing)...

perhaps the solution lies in reserving the "emergency jump" 200 fuel when using flamethrowers and afterburners, while (if hyperdrive jumping to a habited, friendly system with 100 < fuel > 200 from a habited system) accepting that 100 is enough?

**Comment by tehhowch** [comment]
This was very likely solved by recent pulls (namely #2997)

**Comment by Amazinite** [comment]
I'll test it out to see if this is the case, since I always run jump drives with hyperdrives and notice the bug mentioned in this issue often.

**Comment by Amazinite** [comment]
As of f509198, this is still occurring.

**Comment by endless-sky** [comment]
#3123 may have fixed this.

**Comment by oo13** [comment]
I reproduce this issue in 95d7cf5d3c2bf88cd065f4697b0bca670ccb27ba (Current master).

I think this procedure occurs the problem:
1. Ship::Move() set a targetSystem to nullptr when jumping.
2. Entering the next system, nothing set the targetSystem of a escort ship.
3. Ship::JumpRemaining() calculates by the maximum fuel per jump.

This code resolves the problem, but I'm not sure it has no side effect.
```diff
diff --git a/source/Ship.cpp b/source/Ship.cpp
index 30dfde9a..fd888108 100644
--- a/source/Ship.cpp
+++ b/source/Ship.cpp
@@ -2411,7 +2411,14 @@ int Ship::JumpsRemaining() const
 {
        // Make sure this ship has some sort of hyperdrive, and if so return how
        // many jumps it can make.
-       double jumpFuel = JumpFuel(targetSystem);
+       const System *dest = targetSystem;
+       if(!dest)
+       {
+               auto p = GetParent();
+               if(p)
+                       dest = p->GetTargetSystem();
+       }
+       double jumpFuel = JumpFuel(dest);
        return jumpFuel ? fuel / jumpFuel : 0.;
 }
```

**Comment by Pointedstick** [comment]
Looks sane to me. Can't hurt to submit as a PR. Wanna have a crack at it?

**Comment by oo13** [comment]
Thank you.
I'll test it.

**Comment by tehhowch** [comment]
A save which reproduces the issue (depart and perform fleet jumps along the travel plan): 
[jd+hd tester.txt](https://github.com/endless-sky/endless-sky/files/3360150/jd%2Bhd.tester.txt)

On master as of https://github.com/endless-sky/endless-sky/commit/0415035cecab6228ee765ecad0c1581d6ab5c2b4, the escorts will always refuel, even if you are holding J to issue the "let's jump together" command.
With #4288 as of d05df9ab9e011e10bcf924afdf1e331571e5f9af, they will either align for the hyperspace jump, or (if hostiles are present), engage hostile ships. If you jump while they are fighting, they will stop fighting and jump without refueling (as your FS become an out-of-system parent).

The escorts engage hostiles instead of aligning for a jump because they have not yet received a target system for which `Ship::JumpsRemaining` has a non-zero return value, and this is in the behavior checks of AI::Step:
https://github.com/endless-sky/endless-sky/blob/0415035cecab6228ee765ecad0c1581d6ab5c2b4/source/AI.cpp#L820-L821
and was added so that ships that can't jump / can't jump with you can do things like attack enemies instead of frantically trying to get fuel: https://github.com/endless-sky/endless-sky/commit/263ec5724a8f7a3c650840d301cd94822a134df4

I'm not sure if the current state of #4288 is an improvement, as anyone trying to "fleet jump" will instead see their escorts engage hostiles until they scoop up / regen enough fuel to jump, instead of refuel and then prepare to jump (current `master`).

**Comment by Tadrix** [comment]
Personally, in all cases I'm fleet jumping, I'm in a "dangerous part of town", which means I'd want all the escorts refueled ASAP so we can get the hell out of there. Them engaging in fights (and possibly dying while I'm waiting for them) is exactly what I wouldn't want for them to do.

Shooting back is nice and fine, but they shouldn't go out of their way if it means delaying stuff I ordered them to do.

**Comment by tehhowch** [comment]
Merged #4288 since the behavior w.r.t. hostiles in the last jump-from system is not different from the current `master`. If your fleet begins engaging hostiles and you depart, they will now immediately recognize that you have left the system and they can make the jump to your system without needing to refuel. Prior to #4288, they would land to refuel and then jump to join you.
